iree/turbine/kernel/wave/decompose_scan_ops.py

```diff
@@ -116,14 +200,56 @@ def decompose_scan_ops(
             raise NotImplementedError(f"ScanOp '{custom}' not supported")
 
         with custom.graph.inserting_before(custom.fx_node):
-            src, _, _ = node.args
-            assert isinstance(src, fx.Node), f"Scan src is not fx.Node: {type(src)}"
+            scan_src, scan_acc, scan_dim = node.args
+            assert isinstance(
+                scan_src, fx.Node
+            ), f"Scan src is not fx.Node: {type(scan_src)}"
             binary_fn = Add
 
-            result = emit_global_scan(
-                binary_fn, src, custom.graph, subgroup_size, hardware_constraint
+            if scan_dim is None:
+                raise ValueError("No scan dim specified, please specify a scan dim.")
+
+            get_thread_shape = lambda index: max(
+                subs_idxc(x.size) for x in index.values()
             )
 
+            try:
+                op = get_custom(scan_src)
+                thread_shape = get_thread_shape(op.index)
+                local_scan_sizes = thread_shape
+            except Exception as e:
+                index_str = "\n".join(f"{k}: {v}" for k, v in op.index.items())
+                raise RuntimeError(
+                    f"Error in decompose_scan_ops: {scan_src} with index\n"
+                    f"{index_str}\n{scan_src=}\n{scan_acc=}\n{scan_dim=}"
+                ) from e
+
+            if local_scan_sizes > 1:
+                local_scan = emit_local_inclusive_scan(
+                    binary_fn, scan_src, custom.graph, local_scan_sizes
+                )
+                result = emit_global_scan(
+                    binary_fn,
+                    scan_src,
+                    local_scan,
+                    custom.graph,
+                    subgroup_size,
+                    hardware_constraint,
+                    local_scan_sizes,
+                    scan_dim,
+                )
+            else:
+                result = emit_global_scan(
+                    binary_fn,
+                    scan_src,
+                    [scan_src],
+                    custom.graph,
+                    subgroup_size,
+                    hardware_constraint,
+                    local_scan_sizes,
+                    scan_dim,
```

Comment:
IIUC, only `local_sca` here changes, can we take emit_global_scan out of the if condition, and just modify s.t `local_scan` be modified inside the if `local_scan_sizes > 1`